docs/source/developer/threading_implementation.rst

```diff
@@ -0,0 +1,197 @@
+=========================================
+Notes on Numba's threading implementation
+=========================================
+
+The execution of the work presented by the Numba ``parallel`` targets is
+undertaken by the Numba threading layer. Practically, the "threading layer"
+is a Numba built-in library that can perform the required concurrent execution.
+At the time of writing there are three threading layers available, each
+implemented via a different lower level native threading library. More
+information on the threading layers and appropriate selection of a threading
+layer for a given application/system can be found in the
+:ref:`threading layer documentation <numba-threading-layer>`.
+
+The pertinent information to note for the following sections is that the
+function in the threading library that performs the parallel execution is the
+``parallel_for`` function. The job of this function is to both orchestrate and
+execute the parallel tasks.
+
+Thread masking
+--------------
+
+In order to simplify the design, it was decided that Numba should never launch
```

Comment:
So are you saying this restriction of never launching new threads after the initial launch could be removed in the future, assuming someone wants to implement what would be required to make it work?

Reply:
Correct, I think that this restriction could be removed. With thanks to the current design I also think it'd be quite easy to implement (famous last words pending)...

1. Apply restrictions so that only the TBB and OpenMP backends can be used with this behaviour (or rewrite `workqueue` to permit it). I think this reduces to:
   a. Remove restrictions in `set_num_threads()`:
      * Add a "backend" check to be able to check which backend is running
      * If the backend is TBB or OpenMP remove the masking restriction.

   b. I don't think there's anything more to be done. `get_num_threads` should work as-is, the rest of the implementation should be ok.
2. Do something with `_launch_threads()` and the initial threadpool size, this might be "launching" a default size to begin with, such that `NUMBA_NUM_THREADS` simply prescribes the default.
3. Probably worth implementing a `NUMBA_THREAD_LIMIT` or something so as to prevent dynamically creating more threads than cores by default, this can obviously be altered by the user.